radix-engine/src/system/transaction/intent_processor.rs

```diff
@@ -111,27 +115,37 @@ impl<I: TxnInstruction + ManifestDecode + ManifestCategorize> TxnProcessorThread
 
     pub fn resume<Y: SystemApi<RuntimeError> + KernelNodeApi + KernelSubstateApi<L>, L: Default>(
         &mut self,
+        received_value: Option<IndexedScryptoValue>,
         api: &mut Y,
     ) -> Result<ResumeResult, RuntimeError> {
-        while let Some(instruction) = self.instructions.pop_front() {
+        if let Some(received_value) = received_value {
+            self.objects
+                .handle_call_return_data(&received_value, &self.worktop, api)?;
+        }
+
+        while let Some(instruction) = self.remaining_instructions.pop_front() {
             api.update_instruction_index(self.instruction_index)?;
             let (output, yield_instruction) =
                 instruction.execute(&mut self.worktop, &mut self.objects, api)?;
             self.outputs.push(output);
             self.instruction_index += 1;
             if let Some(yield_instruction) = yield_instruction {
                 let result = match yield_instruction {
-                    Yield::ToChild(child) => ResumeResult::YieldToChild(child),
-                    Yield::ToParent => {
-                        // TODO: The below is a David workaround to get tests to pass.
-                        // We should fix this better to have a separate processor for main
-                        // intent and subintents and require subintents to end with a yield to parent.
-                        // They can then return a ResumeResult::DoneYieldToParent or something.
-                        if self.instructions.is_empty() {
-                            // Drop worktop and return ResumeResult::Done
-                            break;
+                    Yield::ToChild(child, value) => ResumeResult::YieldToChild(
+                        child,
+                        IndexedScryptoValue::from_scrypto_value(value),
+                    ),
+                    Yield::ToParent(value) => {
```

Comment:
I'd quite like it if we do some runtime checks here by passing in an `is_subintent` flag, and:
* Checking we only call `Yield::ToParent` in a subintent
* Checking we only return `Ok(ResumeResult::RootIntentDone)` in a non-subintent

Otherwise, returning a RuntimeError. This will catch badly configured `TestTransaction`s which purposefully don't undergo validation.

Reply:
Would prefer that `intent_processor` is not aware of whether it is a child or root. Will add some of these checks in the multithread processor.